sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java

```diff
@@ -0,0 +1,278 @@
+                        discoveryInfo.getVisibility(), taskInfo.getName());
+                continue;
+            }
+            final String directHost;
+            if (isNativeFormat) {
+                // Hostname of agent at offer time:
+                directHost = CommonTaskUtils.getHostname(taskInfo);
+            } else {
+                // Mesos DNS hostname:
+                directHost = String.format("%s.%s.mesos", taskInfo.getName(), serviceName);
+            }
+
+            for (Port port : discoveryInfo.getPorts().getPortsList()) {
+                // host:port to include in 'direct' array(s):
+                final String directHostPort = String.format("%s:%d", directHost, port.getNumber());
+                addPortToEndpoints(
+                        serviceName,
+                        endpointsByName,
+                        taskInfo,
+                        directHostPort,
+                        port.getLabels().getLabelsList());
+            }
+        }
+        return endpointsByName;
+    }
+
+    /**
+     * Adds information about a {@link Port} to the provided {@code endpointsByName}. Information
+     * will be added for any VIPs listed against the {@link Port}'s labels, or if no VIPs are found,
+     * the information will be added against the task type.
+     *
+     * @param endpointsByName the map to write to
+     * @param taskInfo the task which has the port
+     * @param directHostPort the host:port value to advertise for directly connecting to the task
+     * @param portLabels list of any {@link Label}s which were present in the {@link Port}
+     * @throws TaskException if no VIPs were found and the task type couldn't be extracted
+     */
+    private static void addPortToEndpoints(
+            String serviceName,
+            Map<String, JSONObject> endpointsByName,
+            TaskInfo taskInfo,
+            String directHostPort,
+            List<Label> portLabels) throws TaskException {
+        // Search for any VIPs to add the above host:port against:
+        boolean foundAnyVips = false;
+        for (Label label : portLabels) {
+            VIPInfo vipInfo = VIPInfo.parse(taskInfo.getName(), label);
+            if (vipInfo == null) {
+                continue;
+            }
+            // VIP found. file host:port against the VIP name.
+            foundAnyVips = true;
+
+            JSONObject vipEndpoint = endpointsByName.get(vipInfo.name);
+            if (vipEndpoint == null) {
+                vipEndpoint = new JSONObject();
+                endpointsByName.put(vipInfo.name, vipEndpoint);
+            }
+
+            // append entry to 'direct' array for this task:
+            vipEndpoint.append(RESPONSE_KEY_DIRECT, directHostPort);
+            // populate 'vip' field if not yet populated (due to another task with the same vip):
+            vipEndpoint.put(RESPONSE_KEY_VIP, String.format("%s.%s.l4lb.thisdcos.directory:%d",
```

Comment:
Should probably extract the VIP URL format to a constant somewhere for visibility.